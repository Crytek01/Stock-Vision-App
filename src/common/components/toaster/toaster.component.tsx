import { useToaster } from "@/common/contexts/toaster/hook/use-toaster.hook";
import { Snackbar, Alert, SxProps } from "@mui/material";
import { FunctionComponent } from "react";

interface IToaster {}

const ALERT_SX: SxProps = {
  width: "100%",
  padding: "24px",
  fontSize: "1.2rem",
};

export const Toaster: FunctionComponent<IToaster> = () => {
  const { messages } = useToaster();

  return (
    <Snackbar open={true}>
      <div>
        {messages.map(({ id, severity, message }) => (
          <Alert key={id} severity={severity} sx={ALERT_SX}>
            {message}
          </Alert>
        ))}
      </div>
    </Snackbar>
  );
};
