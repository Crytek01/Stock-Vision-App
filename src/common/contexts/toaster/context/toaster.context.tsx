import { createContext } from "react";

interface IToasterContext {
  messages: IInternalToasterMessage[];
  addMessage: (newMessage: IToasterMessage) => void;
}

export const ToasterContext = createContext<IToasterContext | null>(null);
