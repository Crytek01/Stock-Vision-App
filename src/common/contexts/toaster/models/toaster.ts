interface IToasterMessage {
  severity: "success" | "info" | "warning" | "error";
  message: string;
  durationMs?: number;
}

interface IInternalToasterMessage extends IToasterMessage {
  id: string;
}
