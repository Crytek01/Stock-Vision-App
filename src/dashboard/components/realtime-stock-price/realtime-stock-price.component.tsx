import { ISuccessResponse } from "@/common/utils/api/response/api-response";
import { IRealtimeData } from "@/pages/api/stocks";
import { FunctionComponent, useMemo } from "react";
import useSWR from "swr";
import { Line } from "react-chartjs-2";
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from "chart.js";
import { useToaster } from "@/common/contexts/toaster/hook/use-toaster.hook";

ChartJS.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend
);

interface IRealTimeStockPriceProps {
  symbols: string[];
}

const fetcher = async (url: string) =>
  fetch(url)
    .then(
      (response) => response.json() as Promise<ISuccessResponse<IRealtimeData>>
    )
    .then(({ data }) => data);

export const RealtimeStockPrice: FunctionComponent<
  IRealTimeStockPriceProps
> = ({ symbols }) => {
  const { addMessage } = useToaster();

  const symbolsQuery = useMemo(() => {
    return symbols.join(",");
  }, [symbols]);

  //Refreshes every 5 minutes to avoid api limit.
  const { data } = useSWR(`api/stocks?symbols=${symbolsQuery}`, fetcher, {
    refreshInterval: 300000,
    fallbackData: {
      labels: [],
      datasets: [],
    },
    onError: () => {
      addMessage({
        severity: "error",
        message:
          "Failed to retrieve real time stock data from the selected companies.",
        durationMs: 5000,
      });
    },
  });

  return <Line data={data} />;
};
