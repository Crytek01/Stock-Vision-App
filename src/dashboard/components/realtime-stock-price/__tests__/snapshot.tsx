import { render } from "@testing-library/react";
import { RealtimeStockPrice } from "../realtime-stock-price.component";
import { ToasterProvider } from "@/common/contexts/toaster/provider/toaster.context";

it("renders RealtimeStockPrice component", () => {
  const { container } = render(
    <ToasterProvider>
      <RealtimeStockPrice symbols={[]} />
    </ToasterProvider>
  );
  expect(container).toMatchSnapshot();
});
