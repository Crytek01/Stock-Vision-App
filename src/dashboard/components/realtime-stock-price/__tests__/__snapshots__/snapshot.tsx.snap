// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders RealtimeStockPrice component 1`] = `
<div>
  <canvas
    height="150"
    role="img"
    width="300"
  />
</div>
`;
